File: BookStatistics/app/src/main/java/com/github/veselinazatchepina/bookstatistics/abstracts/SingleFragmentAbstractActivity.java
package com.github.veselinazatchepina.bookstatistics.abstracts;


import android.content.Intent;
import android.content.SharedPreferences;
import android.content.res.Configuration;
import android.os.Build;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.annotation.LayoutRes;
import android.support.design.widget.AppBarLayout;
import android.support.design.widget.CollapsingToolbarLayout;
import android.support.design.widget.CoordinatorLayout;
import android.support.design.widget.FloatingActionButton;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.View;

import com.github.veselinazatchepina.bookstatistics.R;
import com.github.veselinazatchepina.bookstatistics.books.activities.AddBookActivity;
import com.github.veselinazatchepina.bookstatistics.utils.AppBarLayoutExpended;
import com.github.veselinazatchepina.bookstatistics.utils.ColorationTextChar;
import com.github.veselinazatchepina.bookstatistics.utils.ThemeUtils;

import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * SingleFragmentAbstractActivity helps avoid boilerplate code.
 */
public abstract class SingleFragmentAbstractActivity extends AppCompatActivity {

    @BindView(R.id.toolbar)
    Toolbar mToolbar;
    @BindView(R.id.appbar_layout)
    AppBarLayout mAppBarLayout;
    @BindView(R.id.collapsing_toolbar_layout)
    CollapsingToolbarLayout mCollapsingToolbarLayout;
    @BindView(R.id.fab)
    FloatingActionButton mFloatingActionButton;

    private SharedPreferences mPrefs;
    private SharedPreferences.OnSharedPreferenceChangeListener mPrefListener;

    public Fragment mCurrentFragment;
    public int mFabImageResourceId;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        definePreferenceListener();
        ThemeUtils.onActivityCreateSetTheme(this);
        defineInputData(savedInstanceState);
        setTitle(ColorationTextChar.setFirstVowelColor(getTitle().toString(), this));
        setContentView(getLayoutResId());
        ButterKnife.bind(this);
        defineToolbar();
        setAppBarNotExpandable();
        setNewTitleStyle();
        defineFragment();
        defineFab();
    }

    private void definePreferenceListener() {
        mPrefs = PreferenceManager.getDefaultSharedPreferences(this);
        mPrefListener = new SharedPreferences.OnSharedPreferenceChangeListener() {
            public void onSharedPreferenceChanged(SharedPreferences prefs, String key) {
                String currentPreferenceThemeName = prefs.getString(key, null);
                ThemeUtils.changeToTheme(SingleFragmentAbstractActivity.this, currentPreferenceThemeName);
            }
        };
    }

    public void defineInputData(Bundle saveInstanceState) {
    }

    @LayoutRes
    protected int getLayoutResId() {
        return R.layout.activity_single_fragment;
    }

    private void defineToolbar() {
        setSupportActionBar(mToolbar);
        if (getSupportActionBar() != null) {
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        }
    }

    private void setAppBarNotExpandable() {
        CoordinatorLayout.LayoutParams layoutParams = (CoordinatorLayout.LayoutParams) mAppBarLayout.getLayoutParams();
        Configuration configuration = getResources().getConfiguration();
        AppBarLayoutExpended.setAppBarLayoutExpended(this, mAppBarLayout, layoutParams, mCollapsingToolbarLayout, configuration);
    }

    public void setNewTitleStyle() {
        setTitle(ColorationTextChar.setFirstVowelColor(getTitle().toString(), this));
    }

    public void defineFragment() {
        FragmentManager fragmentManager = getSupportFragmentManager();
        mCurrentFragment = fragmentManager.findFragmentById(R.id.fragment_container);
        if (mCurrentFragment == null) {
            mCurrentFragment = createFragment();
            fragmentManager.beginTransaction()
                    .add(R.id.fragment_container, mCurrentFragment)
                    .commit();
        }
    }

    public abstract Fragment createFragment();

    private void defineFab() {
        setFabImage();
        mFloatingActionButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                defineActionWhenFabIsPressed();
            }
        });
    }

    private void setFabImage() {
        mFabImageResourceId = setFabImageResourceId();
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            mFloatingActionButton.setImageDrawable(getResources().getDrawable(mFabImageResourceId, getTheme()));
        } else {
            mFloatingActionButton.setImageDrawable(getResources().getDrawable(mFabImageResourceId));
        }
    }

    public int setFabImageResourceId() {
        return R.drawable.ic_add_white_24dp;
    }

    public void defineActionWhenFabIsPressed() {
        Intent intent = AddBookActivity.newIntent(this, getString(R.string.add_book_title));
        startActivity(intent);
    }

    @Override
    protected void onResume() {
        super.onResume();
        mPrefs.registerOnSharedPreferenceChangeListener(mPrefListener);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        mPrefs.unregisterOnSharedPreferenceChangeListener(mPrefListener);
    }
}

File: Books/app/src/main/java/com/github/veselinazatchepina/books/abstracts/SingleFragmentAbstractActivity.kt
package com.github.veselinazatchepina.books.abstracts

import android.content.res.Configuration
import android.os.Bundle
import android.support.design.widget.CoordinatorLayout
import android.support.v4.app.Fragment
import android.support.v4.content.ContextCompat
import android.support.v7.app.AppCompatActivity
import com.github.veselinazatchepina.books.R
import com.github.veselinazatchepina.books.addbook.AddBookActivity
import com.github.veselinazatchepina.books.utils.AppBarLayoutExpanded
import com.github.veselinazatchepina.books.utils.setFirstVowelColor
import kotlinx.android.synthetic.main.activity_single_fragment.*


abstract class SingleFragmentAbstractActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(getLayoutResId())
        defineInputData()
        defineToolbar()
        setAppBarNotExpandable()
        defineNavigationDrawer()
        defineFab()
        if (title != null) {
            setNewTitleStyle(title.toString())
        }
        defineFragment()
    }

    abstract fun createFragment(): Fragment

    open fun getLayoutResId(): Int = R.layout.activity_single_fragment

    open fun defineInputData() {}

    private fun defineToolbar() {
        setSupportActionBar(toolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
    }

    private fun setAppBarNotExpandable() {
        if (resources.configuration.orientation == Configuration.ORIENTATION_LANDSCAPE) {
            AppBarLayoutExpanded.setAppBarLayoutExpended(this,
                    appbarLayout,
                    appbarLayout.layoutParams as CoordinatorLayout.LayoutParams,
                    collapsingToolbarLayout,
                    resources.configuration)
        }
    }

    open fun defineNavigationDrawer() {}

    private fun defineFab() {
        val fabImageResourceId = setFabImageResId()
        addFabButton.setImageDrawable(ContextCompat.getDrawable(this, fabImageResourceId))
        addFabButton.setOnClickListener {
            defineActionWhenFabIsPressed()
        }
    }

    open fun setFabImageResId(): Int = R.drawable.ic_add_white_24dp

    open fun defineActionWhenFabIsPressed() {
        startActivity(AddBookActivity.newIntent(this))
    }

    open fun setNewTitleStyle(title: String) {
        setTitle(title.setFirstVowelColor(this))
    }

    private fun defineFragment() {
        var currentFragment = supportFragmentManager.findFragmentById(R.id.fragmentContainer)
        if (currentFragment == null) {
            currentFragment = createFragment()
            supportFragmentManager.beginTransaction()
                    .add(R.id.fragmentContainer, currentFragment)
                    .commit()
        }
    }
}